File: Longest Substring Without Repeating Characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        int first = 0;
        int last = 0; 
        int max_len = 0;
        unordered_set<char> char_set;

    while (last < s.length())
     {
        if (char_set.find(s[last]) == char_set.end())
         {
            char_set.insert(s[last]);
            last++;
            max_len = max(max_len, last - first);
        } 
        else {
            char_set.erase(s[first]);
            first++;
        }
    }

    return max_len;
    }
};

File: week-2/python/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        i = 0
        j = 0
        max_len = 0
        char_set = set()
  
        while j < len(s):
            if s[j] not in char_set:
                char_set.add(s[j])
                j += 1
                max_len = max(max_len, j - i)
            else:
                char_set.remove(s[i])
                i += 1

        return max_len